File: exercise/Ch5/5-2/5-2.c
#include "../../../lib/tlpi_hdr.h"
#include <unistd.h>
#include <fcntl.h>
#include <assert.h>
#include <sys/stat.h>

int main(int ac, char *av[])
{
	if (ac < 2) {
		usageErr("%s path\n", av[0]);
	}

	/* make sure that the file exists and is not empty */
	struct stat buf;
	if (stat(av[1], &buf) != 0 || buf.st_size == 0) {
		errExit("path must be a non-empty valid file\n");
	}

	int fd = open(av[1], O_WRONLY | O_APPEND);
	if (fd == -1) {
		errExit("open");
	}

	/* Seek to the beginning of the file */
	assert(lseek(fd, 0, SEEK_SET) == 0);

	if (-1 == write(fd, "test", 4)) {
		errExit("write");
	}

	close(fd);
	return EXIT_SUCCESS;
}

File: exercise/Ch5/5-2/5-2.rs
use std::env::args;
use std::fs::{File, OpenOptions};
use std::io::{Seek, SeekFrom, Write};
use std::path::Path;
use std::process::exit;

fn main() {
    let av: Vec<String> = args().collect();
    if av.len() != 2 {
        eprintln!("usage: {} pathname", av[0]);
        exit(1);
    }
    let p: &Path = Path::new(av[1].as_str());
    if !p.exists() || p.metadata().unwrap().len() == 0 {
        eprintln!("path must be a non-empty valid file");
    }

    let mut f: File = OpenOptions::new()
        .write(true)
        .append(true)
        .open(&av[1])
        .expect("can not open or create");

    f.seek(SeekFrom::Start(0)).expect("can not seek");

    f.write("test".as_bytes()).expect("can not write");
}
